SantaFinder.ts

```diff
@@ -1,110 +1,110 @@
-        this.getSantaIds().forEach((santaId) => {
-            const possiblePresentees = this.getPossiblePresenteesById(santaId)
-            this.setPossiblePresentees(santaId, possiblePresentees)
+        this.santas.forEach((santa) => {
+            const possiblePresentees = this.getPossiblePresenteesById(santa.id)
+            santa.possiblePresentees = possiblePresentees
         })
 
-        /* Create possible santa graphs (only full cycles)
-            TODO: Implement multiple cycles? e.g. (a-b-c) (d-e-f) instead of (a-b-c-d-e-f)
-         */
-        const firstSantaId = this.getSantaIds()[0]
-        this.findNextSanta(firstSantaId, this.getSantaById(firstSantaId).possiblePresentees, [])
+        // Go through santas until everyone has a presentee
+        let currentSanta = this.getNextSanta()
+		while (currentSanta) {		
+			this.availableSantas = this.availableSantas.filter((santa) => { return santa != currentSanta.santa })
+			this.availablePresentees = this.availablePresentees.filter((santa) => { return santa != currentSanta.presentee })
+            this.setChosenPresentee(currentSanta.santa, currentSanta.presentee)
+
+            currentSanta = this.getNextSanta()
+		}
+		
+        // Validity check
+        if (this.availableSantas.length || this.availablePresentees.length) {
+            throw 'UNDEFINED-CIRCLE'
+        }
     }
 
-    chooseRandomCycle() {
-        // Choose random cycle
-        const randomIndex = this.getRandomInt(0, this.possibleCycles.length)
-        const chosenCycle = this.possibleCycles[randomIndex]
-        if (!chosenCycle) throw 'UNDEFINED-CIRCLE'
-
-        chosenCycle.forEach((id, i) => {
-            const presenteeId = (i + 1) === chosenCycle.length ? chosenCycle[0] : chosenCycle[i + 1]
-            this.setChosenPresentee(id, presenteeId)
+    private getNextSanta() {
+        // Find all santas that still need an presentee
+        let availableNodes = this.santas
+            .map((santa) => {
+                if (this.availableSantas.includes(santa.id)) {
+                    const presentees = santa.possiblePresentees.filter((presentee: string) => { return this.availablePresentees.includes(presentee) })
+                    return { id: santa.id, possiblePresentees: presentees }
+                }
+            })
+            .filter((item): item is SantaNode => !!item)
+        
+        // We are done
+        if (availableNodes.length === 0) {
+            return null
+        }
+        
+        // Find presentees who should be prioritized because few santas apply
+        const presenteeCounts = availableNodes.flatMap((node) => { return node.possiblePresentees })
+            .reduce((acc, curr) => {
+                return acc[curr] ? ++acc[curr] : acc[curr] = 1, acc
+            }, {} as {[id: string]: number; })
+        const presentees = Object.entries(presenteeCounts)
+            .sort((p1, p2) => { return p1[1] - p2[1] })
+    
```

Comment:
```suggestion
           .sort((p1, p2) => p1[1] - p2[1])
           .map(([presentee]) => presentee)
```
Hier würde ich gleich mappen, weil du ja die Länge nach dem Sortieren nicht mehr brauchst :)